Move hover provider registry to the web worker

twop opened the pull request:
done:
- moved hover registry and all corresponding logic to the webworker

not done:
- **lifetime of hover registry**. The way it worked that registry was always available and reissued hover queries once we have a registered provider. Now though, it is not possible to issue a hover query **before** worker is initialized.
- fix tests
- add tests
- general cleanup.

Please treat this PR as an early wip to validate the direction (for hover registry and all others)

twop commented on browser/src/shared/code-hosts/shared/codeHost.tsx:
```diff
@@ -384,6 +384,7 @@ function initCodeIntelligence({
         getHover: ({ line, character, part, ...rest }) =>
             combineLatest([
+                // TODO what to if a webworker has not been instantiated yet
                 extensionsController.services.textDocumentHover.getHover(
```
this code won't compile because obv there is no hoverRegistry anymore.

twop commented on web/src/backend/features.ts:
```diff
@@ -14,8 +14,13 @@ export function getHover(
+    // extensionsController.services.textDocumentHover.
+    // TODO no longer exists
+    // what to do when there worker has not been created yet?
     return extensionsController.services.textDocumentHover.getHover({
```
this one doesn't compile either. 
Q: is this ok not to have any hovers before worker is instantiated? If not what is desired behavior?

lguychard replied:
Removing services altogether, and directly calling into proxied methods from the worker thread (and thus guaranteeing that the worker has been instantiated before returning an `extensionsController`) will probably be the simplest way of solving this. See example of this here: https://github.com/sourcegraph/sourcegraph/pull/11581

This would solve similar issues with query transformers (["transformers suffer from the same issue: you can issue a transformation before the worker is initialized."](https://sourcegraph.slack.com/archives/CMT39K56Z/p1592450169450500))

felixfbecker replied:
To answer the question, I think that would not be okay. There are basically three options:
1. Ignore hover events while the worker is not instantiated. This would be a bad problem, we already had bugs in the past where when navigating to URLs for a token the hover overlay would not show up because it happened before initialization was done. We need to prevent any flakiness like this.
2. Not render parts of the React app like the Blob component until we have an `extensionController`. This would be bad for page performance.
3. Represent the `extensionController` as a `Promise` or `Observable` in the React component (it can still be passed as an optional prop, since we can represent React props changes as an Observable or Promise). When codeintellify calls `getHover()`, always "await" that `extensionController` first, then call the worker API on a proxy exposed by it (like in @lguychard's PR). Note this will likely not literally use `await` because we need to deal with multiple emissions, so the result would be an `Observable`, but conceptually that is what it would be doing with the `extensionsController`. The end result would essentially be that hovers are "queued up" until the worker is ready (but without having to write an explicit queue anywhere).

twop replied:
It seems that we probably should have a separate PR that utilizes a `Promise<ExtHost>`. And after it this PR would be much simpler

lguychard commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -155,5 +203,54 @@ export const initNewExtensionAPI = (
+    )
+}
+
+/**
+ * adds an element to an array with the ability to remove it back via Unsubscribable.
+ * Both of these changes will be notified via "notifyWith" subject
+ *
+ * @returns Unsubscribable to remove the element from the array.
+ */
+function addElementWithRollback<T>(
+    value: T,
+    {
+        get,
+        set,
+        notifyWith: notify,
+    }: {
+        get: () => T[]
+        set: (val: T[]) => void
+        notifyWith: Subject<T[]>
+    }
+): sourcegraph.Unsubscribable {
+    const modifiedArray = get().concat(value)
+    set(modifiedArray)
+    notify.next(modifiedArray)
+    return {
+        unsubscribe: () => {
+            // eslint-disable-next-line id-length
+            const filtered = get().filter(t => t !== value)
+            set(modifiedArray)
+            notify.next(filtered)
+        },
```
This feels odd to me -- it's a bit like recreating an ad-hoc OOP pattern (encapsulated mutation), except in this case the mutation logic doesn't live next to the state it mutates...

Here's a possible alternative: declare a function in the scope of `initNewExtensionAPI`, that will mutate `extState` _from within the closure_:

```typescript
/**
 * Given a key of ExtState where providers should be stored,
 * returns a provider registration function, and an Observable that will
 * emit the current list of providers upon subscription, and whenever
 * a provider is added or removed.
 */
const withProviders = <TProvider, TKey extends FilterKeys<ExtState, TProvider[]>>(
        key: TKey
    ): [(provider: TProvider) => Unsubscribable, Observable<TProvider[]>] => {
        const providerChanges = new Subject<void>()
        const registerProvider = (provider: TProvider): Unsubscribable => {
            state[key] = [...state[key], provider]
            providerChanges.next()
            return {
                unsubscribe: () => {
                    state[key] = state[key].filter(p => p !== provider)
                    providerChanges.next()
                },
            }
        }
        return [
            registerProvider,
            providerChanges.pipe(
                startWith<void>(undefined),
                map(() => state[key])
            ),
        ]
    }

const [registerHoverProvider, hoverProviderChanges] = withProviders('hoverProviders')
const [registerQueryTransformer, queryTransformerChanges] = withProviders('queryTransformers')
```

twop replied:
I like it more, but the closure over state seems off for me too :(

twop replied:
moved to having `BehaviorSubject` directly in state :(

now that helper seems less awkward
```ts
function addWithRollback<T>(behaviorSubject: BehaviorSubject<T[]>, value: T): sourcegraph.Unsubscribable {
    behaviorSubject.next([...behaviorSubject.value, value])
    return {
        unsubscribe: () => behaviorSubject.next(behaviorSubject.value.filter(p => p != value)),
    }
}

```

twop commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -49,15 +64,24 @@ export type PartialWorkspaceNamespace = Omit<
     // the initial settings data, and instead only subscribe to configuration changes.
     // In order for these extensions to be able to access settings, make sure `configuration` emits on subscription.
 
+    const hoverProvidersChanges = new BehaviorSubject<RegisteredHoverProvider[]>([])
```
I like option 2 a lot actually!

twop replied:
or 1...

twop replied:
I tried to add special "smart" code to have better ergonomics for the option "2" here but I was not happy with it.
In the end I think 
```ts
  const state: ExtState = {
        .... 
        hoverProviders: new BehaviorSubject<RegisteredHoverProvider[]>([]),
    }
```
Is the best in terms of ergonomics (closer to option 1)

twop commented on shared/src/api/extension/hover.ts:
```diff
@@ -48,7 +29,8 @@ export function getHover(
                         provider(parameters).pipe(
-                            finallyReleaseProxy(),
+                            // TODO I think this is not needed anymore
+                            // finallyReleaseProxy(),
```
@felixfbecker do we need to release the result after `wrapRemoteObservable` or somewhere else?
I seems that we don't need to do anything else but wanted to double check

lguychard commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -128,15 +139,15 @@ export const initNewExtensionAPI = (
+                callProviders(
+                    state.hoverProviders,
+                    document,
+                    provider => provider.provideHover(document, toPosition(textParameters.position)),
```
We could do this above, next to `const document` right?

twop replied:
yep

twop commented:
UPD:
- extracted/refactored common logic into `callProviders` function
- added marble tests for that
- added tests for hover merging
- added initial loading value for `getHover` in the main thread (so we have a spinner right away)
- manually tested web app

lguychard commented on shared/src/api/extension/extensionHost.providers.test.ts:
```diff
@@ -0,0 +1,152 @@
+        results => results
+    )
+
+describe('bla', () => {
```
```suggestion
describe('callProviders()', () => {
```

lguychard commented on shared/src/api/extension/extensionHost.hover.test.ts:
```diff
@@ -0,0 +1,120 @@
+        })
+    })
+
+    //
```
```suggestion
```

lguychard commented on shared/src/api/extension/extensionHost.hover.test.ts:
```diff
@@ -0,0 +1,120 @@
+
+    const textHover = (value: string): Hover => ({ contents: { value, kind: MarkupKind.PlainText } })
+
+    describe('integration(ish) tests for scenarios not covered by providers tests', () => {
```
Do we need this additional `describe()` block? I would be OK having this as a comment above the "getHover from ExtensionHost API" describe block above.

lguychard commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -103,6 +134,18 @@ export const initNewExtensionAPI = (
+
+        // Language
+        getHover: (textParameters: TextDocumentPositionParams) => {
+            const document = textDocuments.get(textParameters.textDocument.uri)
+
+            const invokeProvider = (
+                provider: sourcegraph.HoverProvider
+            ): sourcegraph.ProviderResult<sourcegraph.Badged<Hover>> =>
+                provider.provideHover(document, toPosition(textParameters.position))
```
It doesn't seem necessary to create a new `Position` instance every time we invoke a provider?

```suggestion
            const document = textDocuments.get(textParameters.textDocument.uri)
            const position = toPosition(textParameters.position)

            const invokeProvider = (
                provider: sourcegraph.HoverProvider
            ): sourcegraph.ProviderResult<sourcegraph.Badged<Hover>> =>
                provider.provideHover(document, position)
```

lguychard commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -103,6 +134,18 @@ export const initNewExtensionAPI = (
+            ): sourcegraph.ProviderResult<sourcegraph.Badged<Hover>> =>
+                provider.provideHover(document, toPosition(textParameters.position))
+
+            return proxySubscribable(callProviders(state.hoverProviders, document, invokeProvider, mergeHoverResults))
```
With `position` extracted above, could we just have `invokeProvider` declared inline?

```suggestion
            return proxySubscribable(callProviders(state.hoverProviders, document, provider => provider.provideHover(document, position), mergeHoverResults))
```

lguychard commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -155,5 +194,73 @@ export const initNewExtensionAPI = (
     }
 }
+
+// TODO probably worth separate test suit.home
+// maybe copy from registry.ts?
```
Is this TODO still relevant?

twop replied:
it is but probably worth porting tests when the last provider registry is moved to a worker.

twop commented on shared/src/api/extension/flatExtensionApi.ts:
```diff
@@ -20,7 +32,20 @@ export interface ExtState {
+    hoverProviders: BehaviorSubject<RegisteredProvider<sourcegraph.HoverProvider>[]>
+}
+
+interface RegisteredHoverProvider {
+    selector: sourcegraph.DocumentSelector
+    provider: sourcegraph.HoverProvider
+}
```
Right, I think I forgot to change that.

twop commented on web/src/backend/features.ts:
```diff
@@ -14,11 +16,22 @@ export function getHover(
-        },
-    })
+    return concat(
+        [{ isLoading: true, result: null }],
```
I didn't know that this logic exists. Let me check.

twop replied:
I think this is covered by https://sourcegraph.com/github.com/sourcegraph/codeintellify/-/blob/src/hoverifier.ts#L438:17

felixfbecker replied:
Not right there, but close. Found it :)
https://sourcegraph.com/github.com/sourcegraph/codeintellify/-/blob/src/hoverifier.ts#L726:17

twop commented on shared/src/api/extension/extensionHost.providers.test.ts:
```diff
@@ -0,0 +1,152 @@
+                        { uri: 'file:///f.ts' }
+                    )
+                ).toBe('-l-r', {
+                    l: { isLoading: true, result: [LOADING] },
```
If I understand correctly your concern: why we have `loading:true` **and** `result: [LOADING]`?
That is due to in not aggregating results in these tests, we just get values and return them as is
```ts
callProviders(
        providersObservable,
        document,
        value => value,
        results => results // <----------
    )
```
I have tests for aggregation logic in extensionHost.hover.test.ts
```ts
 it('merges non Hover values into nulls', () => {
        expect(mergeHoverResults([LOADING])).toBe(null)
        expect(mergeHoverResults([null])).toBe(null)
        expect(mergeHoverResults([undefined])).toBe(null)
        // and yes, there can be several
        expect(mergeHoverResults([null, LOADING])).toBe(null)
    })
```
So in getHover the final state will be `{isLoading: true, result: null}`

felixfbecker replied:
I feel like filtering out the `LOADING` should be part of `callProviders`. `callProviders` (and `getHover` before it) adds the `LOADING` to the stream basically only internally for each provider so it can then build a `MaybeLoadingResult` of all provider results and their loading states. But now it still builds that `MaybeLoadingResult`, but no longer removes the internal signal from the stream, which feels leaky and makes these tests confusing.

The filtering of `LOADING` will have to be done the same way regardless of whether it's a hover provider, definition, references etc, so it makes sense to me to have `callProviders` do it and not have the `LOADING` emission be part of its public API with every merge function duplicating that filtering (the merge function can focus on merging non-loading results).

twop replied:
That is a good point and I agree. I was hoping to tackle that while migrating the rest of lang providers, once I understand more context about `getHoverLocation`.

twop commented:
> Also want to confirm that this part of the PR description is no longer true, right?

Yes, that works now

twop commented on shared/src/api/extension/extensionHost.hover.test.ts:
```diff
@@ -0,0 +1,120 @@
+    it('merges Hovers with ranges', () => {
+        const hover1: Hover = {
+            contents: { value: 'c1' },
+            // TODO this is weird to cast to ranges
+            range: ({ start: { line: 1, character: 2 }, end: { line: 3, character: 4 } } as unknown) as Range,
```
I totally missed the second constructor for ranges, thanks!

twop replied:
That caused an error in CI, so I reverted back to casting
<img width="901" alt="image" src="https://user-images.githubusercontent.com/6633216/86050684-0c2d7880-ba09-11ea-9a7b-55c6322c3d80.png">

felixfbecker replied:
Ah, it needs to be imported from `@sourcegraph/extension-api-classes`. `sourcegraph` is just the type definition package, it contains no implementations.
